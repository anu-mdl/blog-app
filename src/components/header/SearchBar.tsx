import { Search } from 'lucide-react';
import { Input } from '../ui/input';
import { useState } from 'react';

const SearchBar = ({ isMobile }: { isMobile?: boolean }) => {
  const [searchQuery, setSearchQuery] = useState('');
  const [isSearchFocused, setIsSearchFocused] = useState(false);

  const handleSearch = (query: string) => {
    if (query.trim()) {
      window.location.href = `/blog?search=${encodeURIComponent(query)}`;
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      handleSearch(searchQuery);
      setIsSearchFocused(false);
    }
  };

  return (
    <>
      {isMobile ? (
        <div className="relative">
          <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 h-4 w-4 text-muted-foreground" />
          <Input
            type="search"
            placeholder="Search articles..."
            className="pl-10"
            value={searchQuery}
            onChange={e => setSearchQuery(e.target.value)}
            onKeyDown={handleKeyDown}
          />
        </div>
      ) : (
        <div className="hidden lg:flex flex-1 max-w-md mx-6 relative">
          <div className="relative w-full">
            <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 h-4 w-4 text-muted-foreground" />
            <Input
              type="search"
              placeholder="Search articles..."
              className="pl-10 pr-4"
              value={searchQuery}
              onChange={e => setSearchQuery(e.target.value)}
              onFocus={() => setIsSearchFocused(true)}
              onBlur={() => setTimeout(() => setIsSearchFocused(false), 200)}
              onKeyDown={handleKeyDown}
            />
          </div>
        </div>
      )}
    </>
  );
};

export default SearchBar;
