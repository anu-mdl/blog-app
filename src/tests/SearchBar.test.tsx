import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SearchBar from '@/components/header/SearchBar';

jest.mock('lucide-react', () => ({
  Search: () => <svg data-testid="search-icon" />
}));

jest.mock('@/components/ui/input', () => ({
  Input: ({ children, ...props }: any) => (
    <input {...props} role="textbox">
      {children}
    </input>
  )
}));

const originalLocation = window.location;
beforeAll(() => {
  Object.defineProperty(window, 'location', {
    value: { ...originalLocation, href: '' },
    writable: true
  });
});
afterAll(() => {
  Object.defineProperty(window, 'location', originalLocation);
});

describe('SearchBar', () => {
  const handleLocationMock = jest.fn();

  beforeEach(() => {
    window.location.href = '';
    jest.clearAllMocks();
  });

  it('matches mobile snapshot', () => {
    const { asFragment } = render(<SearchBar isMobile={true} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches desktop snapshot', () => {
    const { asFragment } = render(<SearchBar isMobile={false} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches focused state snapshot', () => {
    const { asFragment } = render(<SearchBar isMobile={false} />);
    fireEvent.focus(screen.getByPlaceholderText('Search articles...'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('updates search query on input', async () => {
    render(<SearchBar isMobile={true} />);
    const input = screen.getByPlaceholderText('Search articles...');
    await userEvent.type(input, 'test query');
    expect(input).toHaveValue('test query');
  });

  it('navigates on enter press with query (mobile)', async () => {
    render(<SearchBar isMobile={true} />);
    const input = screen.getByPlaceholderText('Search articles...');

    await userEvent.type(input, 'react hooks');
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(window.location.href).toContain('/blog?search=react%20hooks');
  });

  it('navigates on enter press with query (desktop)', async () => {
    render(<SearchBar isMobile={false} />);
    const input = screen.getByPlaceholderText('Search articles...');

    await userEvent.type(input, 'next js');
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(window.location.href).toContain('/blog?search=next%20js');
  });

  it('does not navigate with empty query', async () => {
    render(<SearchBar isMobile={false} />);
    const input = screen.getByPlaceholderText('Search articles...');

    fireEvent.keyDown(input, { key: 'Enter' });

    expect(window.location.href).toBe('');
  });

  it('handles focus/blur events (desktop)', async () => {
    render(<SearchBar isMobile={false} />);
    const input = screen.getByPlaceholderText('Search articles...');

    fireEvent.focus(input);
    expect(screen.getByTestId('search-icon').parentElement).toHaveClass(
      'relative'
    ); // Verify focus state

    fireEvent.blur(input);
    setTimeout(() => {
      expect(screen.getByTestId('search-icon').parentElement).not.toHaveClass(
        'relative'
      ); // Verify blur effect
    }, 250);
  });
});
