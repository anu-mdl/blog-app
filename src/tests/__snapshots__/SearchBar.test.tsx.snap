// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar matches desktop snapshot 1`] = `
<DocumentFragment>
  <div
    class="hidden lg:flex flex-1 max-w-md mx-6 relative"
  >
    <div
      class="relative w-full"
    >
      <svg
        data-testid="search-icon"
      />
      <input
        class="pl-10 pr-4"
        placeholder="Search articles..."
        role="textbox"
        type="search"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchBar matches focused state snapshot 1`] = `
<DocumentFragment>
  <div
    class="hidden lg:flex flex-1 max-w-md mx-6 relative"
  >
    <div
      class="relative w-full"
    >
      <svg
        data-testid="search-icon"
      />
      <input
        class="pl-10 pr-4"
        placeholder="Search articles..."
        role="textbox"
        type="search"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchBar matches mobile snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <svg
      data-testid="search-icon"
    />
    <input
      class="pl-10"
      placeholder="Search articles..."
      role="textbox"
      type="search"
      value=""
    />
  </div>
</DocumentFragment>
`;
